pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -0,0 +1,219 @@
+					}
+					nodeValue, ok := node.Labels[constraint.TopologyKey]
+					if !ok {
+						continue
+					}
+					// 6. create a topoPair with key as this TopologySpreadConstraint
+					topoPair := topologyPair{key: constraint.TopologyKey, value: nodeValue}
+					// 7. add the pod with key as this topoPair
+					constraintTopologies[topoPair] = append(constraintTopologies[topoPair], &pod)
+					sumPods++
+				}
+
+				// To actually determine how many pods need to be moved, we sort the topologies in ascending length
+				// [2, 5, 3, 8, 5, 7]
+				//
+				// Would end up like:
+				// [2, 3, 5, 5, 7, 8]
+				//
+				// We then start at i=[0] and j=[len(list)-1] and compare the 2 topology sizes.
+				// If the diff of the size of the topologies is more than the maxSkew, we will move up to half that skew,
+				// or the available pods from the higher topology, whichever number is less.
+				// (Note, we will only move as many pods from a topology as possible without bringing it below the ideal average.)
+				// If the diff is within the skew, we move to the next highest topology.
+				// If the higher topology can't give any more without falling below the average, we move to the next lowest "high" topology
+				//
+				// Following this, the above topologies end up "sorted" as:
+				// [5, 5, 5, 5, 5, 5]
+				// (assuming even distribution by the scheduler of the evicted pods)
+				idealAvg := int(math.Ceil(sumPods / float64(len(constraintTopologies))))
+				sortedTopologies := sortTopologies(constraintTopologies, idealAvg)
+
+				// i is the index for belowOrEqualAvg
+				// j is the index for aboveAvg
+				i := 0
+				j := len(sortedTopologies) - 1
+				for i < j {
+					// if j has no more to give without falling below the ideal average, move to next aboveAvg
+					if len(sortedTopologies[j].pods) <= idealAvg {
+						j--
+					}
+
+					// skew = diff - maxSkew
+					skew := math.Abs(float64(len(sortedTopologies[i].pods) - len(sortedTopologies[j].pods)))
+
+					// if i and j are within the maxSkew of each other, move to next belowOrEqualAvg
+					if int32(skew) <= constraint.MaxSkew {
+						i++
+						continue
+					}
+
+					// the most that can be given from aboveAvg is up to halfway,
+					// or how many it has remaining without falling below the average
+					movePods := int(math.Min(math.Ceil(skew/2), float64(len(sortedTopologies[j].pods)-idealAvg)))
+					if movePods == 0 {
+						i++
+						continue
+					}
+
+					// remove pods from aboveAvg and add them to the list of pods to be evicted
+					// also (just for tracking), add them to the list of pods in belowOrEqualAvg[i]
+					aboveToEvict := sortedTopologies[j].pods[movePods-1:]
+					for _, pod := range aboveToEvict {
+						podsForEviction[pod] = struct{}{}
+					}
+					sortedTopologies[j].pods = sortedTopologies[j].pods[:movePods]
+					sortedTopologies[i].pods = append(sortedTopologies[i].pods, aboveToEvict...)
```

Comment:
What do these 2 lines mean?

Reply:
`sortedTopologies[j].pods = sortedTopologies[j].pods[:movePods]` - shortens the list of available pods to be moved from the "higher" domain

`sortedTopologies[i].pods = append(sortedTopologies[i].pods, aboveToEvict...)` adds the evicted pods to the "lower" domain

I meant this to keep track of what pods can still be moved (so we don't mark the same pod twice) and track how many still need to be moved by updating the smaller domain's list with a simulated ideal movement